---
title: genBarcode_LARRY
output: html_notebook
---

Load libraries:

```{r}
library(genBaRcode)
library(dplyr)
library(tidyr)
library(ggplot2)
```

1.  **processingRawData():**

    translation of raw next generation sequencing (NGS) data files into a sorted table of BC sequences and its corresponding read counts


**LARRY BARCODE:**

EGFP side,close to primer20: TGNNNNTGNNNNCANNNNACNNNNGANNNNGTNNNNAGNNNN (removed AT because maybe not in all sequences)

END, other side: ATNNNNCTNNNNACNNNNTCNNNNGTNNNNTGNNNNCANNNNCA

Set variables for arguments:

```{r}
directory <- "/Users/boeziog/Desktop/LARRY/Library_diversity_check/src"

file_input <- "Plasmid_BOE4508A2_S2_L001_R2_001_sub.fastq"

output_label <- "LARRY_plasmid_sample2_R2"

res_dir <- "/Users/boeziog/Desktop/LARRY/Library_diversity_check/results"

plots_dir <- "/Users/boeziog/Desktop/LARRY/Library_diversity_check/results/plots"

bb <- "TGNNNNTGNNNNCANNNNACNNNNGANNNNGTNNNNAGNNNN"

```


Run function: 

```{r}
BC_dat <- processingRawData(file_name = file_input, 
                            source_dir = directory, results_dir = res_dir, 
                            mismatch = 1, label = output_label, 
                            bc_backbone = bb, min_reads = 3, min_score = 30,
                            save_it = TRUE, seqLogo = FALSE, full_output = FALSE, 
                            wobble_extraction = FALSE)


nrow(BC_dat@reads)

```


```{r}

show(BC_dat)

head(getReads(BC_dat))
```


**2. Error correction (EC):** error-correction (EC) of the detected BCs.

BC_dat=BC_data, maxDist=4, save_it=FALSE, cpus=1, strategy="sequential", m="hamming", type="clustering", only_EC_BCs=TRUE, EC_analysis=FALSE, start_small=FALSE


```{r}
BC_dat_EC_4 <- errorCorrection(BC_dat, maxDist = 4, save_it = TRUE, 
                              m = "hamming", type = "standard", EC_analysis = FALSE)

show(BC_dat_EC_4)

```
```{r}
nrow(BC_dat_EC_4@reads)
```



**3. Graphic visualization**

**3a: Quality scores**

Plot quality score of reads in histogram:
```{r}
plotQualityScoreDis(source_dir = directory, file_name = file_input, type = "mean")

ggsave(path = plots_dir,
  filename = paste0(output_label, "_QC.pdf"),
  plot = last_plot(),
)
```


**3a.1: Plot nt distribution across all reads**

```{r}
plotSeqLogo(BC_dat=BC_dat_EC_4,colrs=NULL)

ggsave(path = plots_dir,
  filename = paste0(output_label,"_seqlogo.pdf"),
  plot = last_plot(),
  bg = "transparent"
)
```

**3b: Plotting Barcode Frequencies**

```{r}
generateKirchenplot (BC_dat_EC_4, ori_BCs = NULL, ori_BCs2 = NULL, loga = TRUE, 
                     col_type = NULL, m = "hamming")

ggsave(path = plots_dir,
  filename = paste0(output_label,"_kirchenplot.png"),
  plot = last_plot(),
  bg = "transparent",
)
```

Plot barcode distances:
```{r}
ggplotDistanceGraph(BC_dat=BC_dat_EC_4,minDist=1,loga=TRUE,m="hamming")

ggsave(path = plots_dir,
  filename = paste0(output_label,"_distgraph.pdf"),
  plot = last_plot(),
  bg = "transparent",
)
```

